File: src/taigi_converter/converter.py
import re
import unicodedata

from taigi_converter.phonetics import is_stop_tone, parse_syllable, strip_tone_mark
from taigi_converter.poj import to_poj
from taigi_converter.tables import nfc
from taigi_converter.tl import to_tl
from taigi_converter.zhuyin import to_zhuyin

_SYSTEMS = {"tl", "poj", "zhuyin"}

_SYLLABLE_RE = re.compile(
    r"([A-Za-z\u0300-\u030d\u030b\u0306\u0358\u207f"
    r"\u00e1\u00e0\u00e2\u00ea\u00e9\u00e8\u00f3\u00f2\u00f4\u00fa\u00f9\u00ec\u00ee"
    r"\u0101\u0113\u012b\u014d\u016b\u01ce\u011b\u030d"
    r"]+[0-9]?)"
)


def convert(text, source, target):
    if source not in _SYSTEMS:
        raise ValueError(f"Unknown source system: {source}")
    if target not in _SYSTEMS:
        raise ValueError(f"Unknown target system: {target}")

    if source == target:
        return text

    if source == "zhuyin":
        raise ValueError("Zhuyin as source is not yet supported")

    if target == "zhuyin":
        return _convert_to_zhuyin(text, source)

    return _convert_romanization(text, source, target)


def _convert_to_zhuyin(text, source):
    numbered = to_tone_number(text)
    lines = numbered.splitlines()
    result_lines = []
    for line in lines:
        tokens = line.split(" ")
        converted = "".join(to_zhuyin(t) for t in tokens)
        result_lines.append(converted)
    return "\n".join(result_lines)


def _convert_romanization(text, source, target):
    assembler = to_tl if target == "tl" else to_poj

    def replace_syllable(match):
        syllable = match.group(1)
        try:
            initial, final, tone = parse_syllable(syllable)
        except ValueError:
            return syllable
        case = _detect_case(syllable)
        result = assembler(initial, final, tone)
        return _apply_case(result, case)

    return _SYLLABLE_RE.sub(replace_syllable, text)


def to_tone_number(text):
    decomposed = unicodedata.normalize("NFD", text)
    result = []
    i = 0
    while i < len(decomposed):
        if decomposed[i].isalpha() or decomposed[i] in "\u0358\u207f":
            syllable_start = i
            while i < len(decomposed) and (
                decomposed[i].isalpha()
                or unicodedata.category(decomposed[i]).startswith("M")
                or decomposed[i] in "\u0358\u207f"
            ):
                i += 1
            if i < len(decomposed) and decomposed[i].isdigit():
                result.append(decomposed[syllable_start : i + 1])
                i += 1
                continue
            chunk = decomposed[syllable_start:i]
            bare, tone = strip_tone_mark(unicodedata.normalize("NFC", chunk))
            if tone:
                result.append(nfc(bare) + tone)
            else:
                bare_nfc = nfc(bare)
                try:
                    _, final, _ = parse_syllable(bare_nfc)
                    if is_stop_tone(final):
                        result.append(bare_nfc + "4")
                    else:
                        result.append(bare_nfc + "1")
                except ValueError:
                    result.append(bare_nfc)
        else:
            result.append(decomposed[i])
            i += 1
    return "".join(result)


def to_tone_mark(text, system="tl"):
    assembler = to_tl if system == "tl" else to_poj

    def replace_match(match):
        syllable = match.group(1)
        if not re.search(r"[0-9]$", syllable):
            return syllable
        try:
            initial, final, tone = parse_syllable(syllable)
        except ValueError:
            return syllable
        case = _detect_case(syllable)
        result = assembler(initial, final, tone)
        return _apply_case(result, case)

    return _SYLLABLE_RE.sub(replace_match, text)


def _detect_case(text):
    alpha = "".join(c for c in text if c.isalpha())
    if not alpha:
        return "lower"
    if alpha.isupper():
        return "upper"
    if alpha[0].isupper():
        return "title"
    return "lower"


def _apply_case(text, case):
    if case == "upper":
        return text.upper()
    if case == "title":
        return text[0].upper() + text[1:] if text else text
    return text


class TailoTpsConverter:
    def convert(self, text, *, encode_safe=False):
        lines = text.splitlines()
        result_lines = []
        for line in lines:
            tokens = line.split(" ")
            converted = "".join(to_zhuyin(t, encode_safe=encode_safe) for t in tokens)
            result_lines.append(converted)
        return "\n".join(result_lines)

    def convert_file(self, input_path, output_path, *, encode_safe=False):
        with open(input_path, encoding="utf-8") as f:
            text = f.read()
        result = self.convert(text, encode_safe=encode_safe)
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(result)
        return result

File: src/converter.js
import { isStopTone, normalizeToTl, parseSyllable, splitInitialFinal, stripToneMark } from "./phonetics.js";
import { toTl } from "./tl.js";
import { toPoj } from "./poj.js";
import { fromZhuyin, toZhuyin } from "./zhuyin.js";
import { segmentWords } from "./segmenter.js";

const SYSTEMS = new Set(["tl", "poj", "zhuyin"]);

// Match TL/POJ syllables in NFC. Class covers:
//   - ASCII A-Za-z
//   - Latin-1 Supplement U+00C0-U+00FF (lower + UPPER precomposed acute/grave/
//     circumflex on a/e/i/o/u — `á à â ê é è í ì î ó ò ô ú ù û` and uppers)
//   - Latin Extended-A U+0100-U+017F (macrons + breves on a/e/i/o/u —
//     `ā ē ī ō ū` and `ă ĕ ĭ ŏ ŭ` POJ tone 9 plus uppers)
//   - Latin Extended-B U+01CD-U+01DC (carons `ǎ ě ǐ ǒ ǔ` for tone 6 plus uppers)
//   - Combining diacritics U+0300-U+036F (acute/grave/circumflex/macron/caron/
//     breve/double-acute/vertical-line/combining-dot — covers all decomposed
//     forms inc. tone 8 U+030D, TL tone 9 U+030B, combining `oo` dot U+0358)
//   - POJ nasal markers ⁿ (U+207F) and ᴺ (U+1D3A)
// The earlier narrower class missed UPPERCASE precomposed + several lowercase
// forms (`í û ǐ ǒ ǔ ă`), causing `convert` to silently bypass valid inputs
// like `Ká` or `kă`.
const SYLLABLE_RE = new RegExp(
  "([A-Za-z\\u00C0-\\u00FF\\u0100-\\u017F\\u01CD-\\u01DC" +
  "\\u0300-\\u036F\\u207F\\u1D3A" +
  "]+[0-9]?)", "g"
);

export function convert(text, source, target) {
  if (!SYSTEMS.has(source)) throw new Error(`Unknown source system: ${source}`);
  if (!SYSTEMS.has(target)) throw new Error(`Unknown target system: ${target}`);
  if (source === target) return text;
  if (source === "zhuyin") {
    const numbered = segmentWords(fromZhuyin(text));
    return toToneMark(numbered, target);
  }

  if (target === "zhuyin") {
    const numbered = toToneNumber(text);
    return numbered.split("\n").map(line => {
      const parts = [];
      line.split(" ").forEach(word => {
        const prefix = word.startsWith("--") ? "--" : "";
        const bare = prefix ? word.slice(2) : word;
        bare.split("-").forEach((t, i) => {
          if (!t) return;
          const tps = toZhuyin((i === 0 ? prefix : "") + t).trimEnd();
          tps.split(/([。，？．「」]+)/).forEach(s => {
            if (s) parts.push(s);
          });
        });
      });
      return parts.join(" ");
    }).join("\n");
  }

  const assembler = target === "tl" ? toTl : toPoj;
  return text.replace(SYLLABLE_RE, (match) => {
    try {
      const [initial, final, tone] = parseSyllable(match);
      const caseType = detectCase(match);
      const result = assembler(initial, final, tone);
      return applyCase(result, caseType);
    } catch {
      return match;
    }
  });
}

export function toToneNumber(text) {
  const decomposed = text.normalize("NFD");
  const result = [];
  let i = 0;
  while (i < decomposed.length) {
    const ch = decomposed[i];
    if (/\p{L}/u.test(ch) || ch === "\u0358" || ch === "\u207f" || ch === "\u1d3a") {
      const syllableStart = i;
      while (i < decomposed.length && (
        /\p{L}/u.test(decomposed[i]) ||
        /^\p{M}/u.test(decomposed[i]) ||
        decomposed[i] === "\u0358" || decomposed[i] === "\u207f" || decomposed[i] === "\u1d3a"
      )) { i++; }
      if (i < decomposed.length && /[0-9]/.test(decomposed[i])) {
        result.push(decomposed.slice(syllableStart, i + 1));
        i++;
        continue;
      }
      const chunk = decomposed.slice(syllableStart, i);
      const nfcChunk = chunk.normalize("NFC");
      const [bare, tone] = stripToneMark(nfcChunk);
      if (tone) {
        result.push(bare.normalize("NFC") + tone);
      } else {
        const bareNfc = bare.normalize("NFC");
        try {
          const [, final] = splitInitialFinal(normalizeToTl(bareNfc.toLowerCase()));
          result.push(bareNfc + (isStopTone(final) ? "4" : "1"));
        } catch {
          result.push(bareNfc);
        }
      }
    } else {
      result.push(decomposed[i]);
      i++;
    }
  }
  return result.join("");
}

export function toToneNumberAscii(text) {
  return toToneNumber(text)
    .replaceAll("o͘", "oo")
    .replaceAll("O͘", "OO")
    .replaceAll("ⁿ", "nn")
    .replaceAll("ᴺ", "NN");
}

export function toToneMark(text, system = "tl") {
  const assembler = system === "tl" ? toTl : toPoj;
  return text.replace(SYLLABLE_RE, (match) => {
    if (!/[0-9]$/.test(match)) return match;
    try {
      const [initial, final, tone] = parseSyllable(match);
      const caseType = detectCase(match);
      const result = assembler(initial, final, tone);
      return applyCase(result, caseType);
    } catch {
      return match;
    }
  });
}

function detectCase(text) {
  const alpha = [...text].filter(c => /\p{L}/u.test(c)).join("");
  if (!alpha) return "lower";
  if (alpha === alpha.toUpperCase()) return "upper";
  if (alpha[0] === alpha[0].toUpperCase()) return "title";
  return "lower";
}

function applyCase(text, caseType) {
  if (caseType === "upper") return text.toUpperCase().replaceAll("\u207f", "\u1d3a");
  if (caseType === "title") return text ? text[0].toUpperCase() + text.slice(1) : text;
  return text;
}
